### PR TITLE
Sidekick improvements

### PR DESCRIPTION
**Pull request by sasa1977**
This is a PR on top of your PR :-)

I went on to try your changes locally, and I've encountered some issues which I decided to fix In the process I also made various refactorings and some other changes. 

There are some other changes I'd like to try out, but I won't find the time to do this today, so I'm submitting what I have now, and we can do the rest in a separate PR.

The PR is split into smaller self-contained commits, so I suggest reviewing one commit at a time, rather than looking at the complete diff. Let me know what you think. If you're ok with these changes, we can merge this into your branch, and then your branch into docker.

### CODE REVIEW

**Comment by sasa1977** [comment]
Hey, so I pushed a couple of more changes, but I think I'm going to stop for now :-)

Assuming you agree with these changes, here are some near-future changes I'd like to see:

1. Remove default `:docker` name. Force the client to pick an explicit name instead.
2. Expand tests. I added a very simple test for the moment, mainly to avoid the need to fire up `iex` and start things manually. We could already test a couple of more things, like e.g. if test process is started.
3. Link the sidekick supervisor to the caller process instead of monitoring the parent node.
4. Wrap the port inside a GenServer, which exits if the sidekick node stops. The GenServer could also debug log any output obtained from the sidekick node.
5. Add a custom logger backend on the sidekick node which forwards log messages to the parent.

I think 1 and 2 are fairly small changes that could be added to this PR (or rather your PR after this one gets merged). 3, 4, and 5 are probably best done in separate PRs. Perhaps 3 and 4 could be done in a single PR, where 4 is done first, and then 3 starts making more sense (since port is at that point started from a long-running process).

But anyway, I propose we wrap up this PR, then your PR, and then we can move on to other improvements. Let me know what you think.

File: lib/sidekick/sidekick.ex [hunk 1]
```diff
@@ -1,75 +1,88 @@
...
     boot_file_args = "-boot #{priv_dir}/node"
 
     cookie = Node.get_cookie()
     cookie_arg = "-setcookie #{cookie}"
 
-    paths_arg = "-pa #{paths}"
+    paths_args = :code.get_path() |> Enum.map(&"-pa #{&1}") |> Enum.join(" ")
 
-    command_args = "-s Elixir.Sidekick start_sidekick #{parent_node}"
+    encoded_arg = {self(), children} |> :erlang.term_to_binary() |> Base.encode32(padding: false)
```
Comment (Hanspagh) [inline lib/sidekick/sidekick.ex:82]: Ahh sweet, I had such a hard time figuring out how to do this. That I moved the init logic back into start node. This is much metter

File: lib/sidekick/sidekick.ex [hunk 2]
```diff
@@ -1,75 +1,88 @@
...
   end
 
-  defp call(node, module, function, args) do
-    :rpc.block_call(node, module, function, args)
+  defp ensure_distributed! do
+    System.cmd("epmd", ["-daemon"])
+
+    node_name = :crypto.strong_rand_bytes(16) |> Base.encode32(padding: false, case: :lower)
+
+    case :net_kernel.start([:"#{node_name}@127.0.0.1"]) do
```
Comment (Hanspagh) [inline lib/sidekick/sidekick.ex:18]: Should we hardcode hostname here?

**Review by Hanspagh — APPROVED** [review]
I think these are some great improvements to readability and overall stability.

**Comment by Hanspagh** [comment]
There is one test failing, but it does not look like it is related to this, should I just merge. And then continue on point 1 and two from above?

File: lib/sidekick/sidekick.ex (hunk 2 above)
Comment (sasa1977) [reply lib/sidekick/sidekick.ex:18]: This is a convenience auto-switch to distribution, which is needed only in dev/test. I think 127.0.0.1 is a reasonable hardcoded choice that should work on all platforms, but I'm open to counter arguments :-)

**Comment by sasa1977** [comment]
> There is one test failing, but it does not look like it is related to this, should I just merge. And then continue on point 1 and two from above?

Yeah, this test is unrelated. I'll take a look at it separately. If you're fine with the changes feel free to merge and resume with points 1 and 2, and then I think we can merge your PR and do the rest in separate PRs.

**Pull request merged**